### PR TITLE
SIG-1752 Validate postback checksum

### PR DESCRIPTION
**Pull request by Nuss9**
The method IsPostbackChecksumValid accepts a postback and shared secret, both as strings. 
The parameters Transaction Id and Status will be taken from the postback, and together with the given shared secret a formula will calculate the checksum. When this calculated checksum matches with the postback's checksum, the method will return true, all other outcomes will result in false.

### CODE REVIEW

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 1]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+                x.ShouldThrow<ErrorHandling.BadAuthorizationException>();
+            }
+        }
+
+
 		[Fact]
 		public async void when_AddOrReplaceFileMetaToTransaction_is_called_then_the_request_body_should_contain_the_serialized_file_meta()
 		{
 			var mockHttp = new MockHttpMessageHandler();
-
```
Comment (WouterJanson) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Should not be part of commit

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 2]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+
+                //Assert
+                mockHttp.VerifyNoOutstandingRequest();
+                x.ShouldNotThrow<ErrorHandling.BadAuthorizationException>();
+            }
+        }
+
+        [Fact]
+        public void calculated_invalid_checksum_results_in_BadAuthorizationException() {
+
```
Comment (WouterJanson) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: No newline

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 3]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+        [Fact]
+        public void calculated_valid_checksum_results_in_no_exception() {
+            //Arrange
+            var mockHttp = new MockHttpMessageHandler();
+            var validChecksum = "0f628893e173dce50c89238493e3a21a61267b21";
+            var transactionId = "81eef8bf-0661-4f6f-aa2f-5cb506cdc065";
+            var transactionStatus = "10";
+            var sharedSecret = "UwMKSuFPtydaZHpNzVyJDfgX7f1wMf6F";
+
+
```
Comment (WouterJanson) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: One newline too many

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 4]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+        public void calculated_invalid_checksum_results_in_BadAuthorizationException() {
+        
+            //Arrange
+            var mockHttp = new MockHttpMessageHandler();
+            var validChecksum = "1f628893e173dce50c89238493e3a21a61267b21";
+            var transactionId = "81eef8bf-0661-4f6f-aa2f-5cb506cdc065";
+            var transactionStatus = "10";
+            var sharedSecret = "UwMKSuFPtydaZHpNzVyJDfgX7f1wMf6F";
+
+
```
Comment (WouterJanson) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: One newline too many

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 5]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
 			Endpoint = "http://localhost/api/"
 		};
 
+        [Fact]
+        public void calculated_valid_checksum_results_in_no_exception() {
```
Comment (WouterJanson) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Opening bracket for method should be on a new line

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 6]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+            //Arrange
+            var mockHttp = new MockHttpMessageHandler();
+            var validChecksum = "0f628893e173dce50c89238493e3a21a61267b21";
+            var transactionId = "81eef8bf-0661-4f6f-aa2f-5cb506cdc065";
+            var transactionStatus = "10";
+            var sharedSecret = "UwMKSuFPtydaZHpNzVyJDfgX7f1wMf6F";
+
+
+            using (var httpClient = mockHttp.ToHttpClient())
+            {
```
Comment (WouterJanson) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Opening bracket for using should not be on a newline

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 7]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+                Action x = () => signhostApiClient.CalculateAndValidateChecksum(transactionId, transactionStatus, sharedSecret, validChecksum);
+
+                //Assert
+                mockHttp.VerifyNoOutstandingRequest();
+                x.ShouldNotThrow<ErrorHandling.BadAuthorizationException>();
+            }
+        }
+
+        [Fact]
+        public void calculated_invalid_checksum_results_in_BadAuthorizationException() {
```
Comment (WouterJanson) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Opening bracket for method should be on a new line

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 8]
```diff
@@ -21,6 +21,32 @@ public class SignHostApiClient
...
+        /// Calculates and validates checksum.
+        /// </summary>
+        /// <param name="transactionId">Transaction identifier.</param>
+        /// <param name="status">Status.</param>
+        /// <param name="sharedSecret">Shared secret.</param>
+        public void CalculateAndValidateChecksum(                               //has to become private
+                                        string transactionId,
+                                        string transactionStatus,
+                                        string sharedSecret,
+                                        string validChecksum) {
```
Comment (WouterJanson) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Opening bracket for method should be on a new line

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 9]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+            //Arrange
+            var mockHttp = new MockHttpMessageHandler();
+            var validChecksum = "1f628893e173dce50c89238493e3a21a61267b21";
+            var transactionId = "81eef8bf-0661-4f6f-aa2f-5cb506cdc065";
+            var transactionStatus = "10";
+            var sharedSecret = "UwMKSuFPtydaZHpNzVyJDfgX7f1wMf6F";
+
+
+            using (var httpClient = mockHttp.ToHttpClient())
+            {
```
Comment (WouterJanson) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Opening bracket for using should not be on a newline

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 10]
```diff
@@ -21,6 +21,32 @@ public class SignHostApiClient
 		private readonly ISignHostApiClientSettings settings;
 		private readonly HttpClient client;
 
+        /// <summary>
+        /// Calculates and validates checksum.
+        /// </summary>
+        /// <param name="transactionId">Transaction identifier.</param>
+        /// <param name="status">Status.</param>
+        /// <param name="sharedSecret">Shared secret.</param>
```
Comment (WouterJanson) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: parameter documentation for `validChecksum` is missing

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs (hunk 2 above)
Comment (MrJoe) [reply src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Additionally the `{` should be on its own line here.

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 11]
```diff
@@ -21,6 +21,32 @@ public class SignHostApiClient
...
+            string calculatedChecksum;
+            using (var sha1 = System.Security.Cryptography.SHA1.Create()) {
+                var x = $"{transactionId}||{transactionStatus}|{sharedSecret}";
+                var byteArray = System.Text.Encoding.UTF8.GetBytes(x);
+                byte[] hash = sha1.ComputeHash(byteArray);
+                calculatedChecksum = System.BitConverter
+                                           .ToString(hash)
+                                           .Replace("-", string.Empty)
+                                           .ToLower();
+            };
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: A closing `}` should not have a ;,
After a closing `}` there should be a empty newline (unless followed by another `}`'

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 12]
```diff
@@ -21,6 +21,32 @@ public class SignHostApiClient
...
+                                        string validChecksum) {
+            string calculatedChecksum;
+            using (var sha1 = System.Security.Cryptography.SHA1.Create()) {
+                var x = $"{transactionId}||{transactionStatus}|{sharedSecret}";
+                var byteArray = System.Text.Encoding.UTF8.GetBytes(x);
+                byte[] hash = sha1.ComputeHash(byteArray);
+                calculatedChecksum = System.BitConverter
+                                           .ToString(hash)
+                                           .Replace("-", string.Empty)
+                                           .ToLower();
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Unneeded string operations, try to reduce to ease the load on the GC

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 13]
```diff
@@ -21,6 +21,32 @@ public class SignHostApiClient
...
+        /// <param name="status">Status.</param>
+        /// <param name="sharedSecret">Shared secret.</param>
+        public void CalculateAndValidateChecksum(                               //has to become private
+                                        string transactionId,
+                                        string transactionStatus,
+                                        string sharedSecret,
+                                        string validChecksum) {
+            string calculatedChecksum;
+            using (var sha1 = System.Security.Cryptography.SHA1.Create()) {
+                var x = $"{transactionId}||{transactionStatus}|{sharedSecret}";
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Variable `x` doesn't mean anything to me, give them better naming to describe what the variable is describing.

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 14]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+                var signhostApiClient = new SignHostApiClient(settings, httpClient);
+
+                //Act
+                Action x = () => signhostApiClient.CalculateAndValidateChecksum(transactionId, transactionStatus, sharedSecret, validChecksum);
+
+                //Assert
```
Comment (WouterHD) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Please leave a space between slashes and the comment, like `// Assert`

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 15]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+            var transactionId = "81eef8bf-0661-4f6f-aa2f-5cb506cdc065";
+            var transactionStatus = "10";
+            var sharedSecret = "UwMKSuFPtydaZHpNzVyJDfgX7f1wMf6F";
+
+
+            using (var httpClient = mockHttp.ToHttpClient())
+            {
+                var signhostApiClient = new SignHostApiClient(settings, httpClient);
+
+                //Act
```
Comment (WouterHD) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Please leave a space between slashes and the comment, like `// Act`

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 16]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
 			Endpoint = "http://localhost/api/"
 		};
 
+        [Fact]
+        public void calculated_valid_checksum_results_in_no_exception() {
+            //Arrange
```
Comment (WouterHD) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Please leave a space between slashes and the comment, like `// Arrange`

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 17]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+            var transactionId = "81eef8bf-0661-4f6f-aa2f-5cb506cdc065";
+            var transactionStatus = "10";
+            var sharedSecret = "UwMKSuFPtydaZHpNzVyJDfgX7f1wMf6F";
+
+
+            using (var httpClient = mockHttp.ToHttpClient())
+            {
+                var signhostApiClient = new SignHostApiClient(settings, httpClient);
+
+                //Act
```
Comment (WouterHD) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Please leave a space between slashes and the comment

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 18]
```diff
@@ -19,11 +20,58 @@ public class SignHostApiClientTests
...
+                Action x = () => signhostApiClient.CalculateAndValidateChecksum(transactionId, transactionStatus, sharedSecret, validChecksum);
+
+                //Assert
```
Comment (WouterHD) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Please leave a space between slashes and the comment

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs (hunk 12 above)
Comment (Nuss9) [reply src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Explanation required.

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 19]
```diff
@@ -19,11 +19,53 @@ public class SignHostApiClientTests
...
+			validatePostback.ShouldBeEquivalentTo(false);
+		}
+
+		[Fact]
+		public void when_IsPostbackChecksumValid_is_called_with_invalid_postback_format_then_exception_is_thrown()
+		{
+			// Arrange
+			string sharedSecret = "UwMKSuFPtydaZHpNzVyJDfgX7f1wMf6F";
+			string falsepostback = @"{""I'm not a valid postback"":""I'm still not a valid postback""}";
+			//string falsepostback = "I'm not a valid postback";
```
Comment (WouterHD) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Don't commit commented code

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 20]
```diff
@@ -19,11 +19,53 @@ public class SignHostApiClientTests
...
+
+			// Act
+			Action validatePostback = () => SignHostApiClient.IsPostbackChecksumValid(falsepostback, sharedSecret);
```
Comment (WouterHD) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Using `Action` is not consistent with other tests, if the format is invalid shouldn't the `IsPostbackChecksumValid` just return `false`?

File: src/SignhostAPIClient/Rest/ErrorHandling/HttpResponseMessageErrorHandlingExtensions.cs [hunk 21]
```diff
@@ -78,7 +79,7 @@ public static class HttpResponseMessageErrorHandlingExtensions
 						errorMessage);
 			}
 
-			System.Diagnostics.Debug.Fail("Should not be reached");
+			Debug.Fail("Should not be reached");
```
Comment (WouterHD) [inline src/SignhostAPIClient/Rest/ErrorHandling/HttpResponseMessageErrorHandlingExtensions.cs]: This change should not be part of this commit

File: src/SignhostAPIClient/Rest/ErrorHandling/HttpResponseMessageErrorHandlingExtensions.cs [hunk 22]
```diff
@@ -62,7 +63,7 @@ public static class HttpResponseMessageErrorHandlingExtensions
 
 			switch (response.StatusCode) {
 				case HttpStatusCode.Unauthorized:
-					throw new System.UnauthorizedAccessException(
+					throw new UnauthorizedAccessException(
```
Comment (WouterHD) [inline src/SignhostAPIClient/Rest/ErrorHandling/HttpResponseMessageErrorHandlingExtensions.cs]: This change should not be part of this commit

File: src/SignhostAPIClient/Rest/ErrorHandling/HttpResponseMessageErrorHandlingExtensions.cs [hunk 23]
```diff
@@ -1,4 +1,5 @@
 using System;
+using System.Diagnostics;
```
Comment (WouterHD) [inline src/SignhostAPIClient/Rest/ErrorHandling/HttpResponseMessageErrorHandlingExtensions.cs]: This change should not be part of this commit

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 24]
```diff
@@ -21,6 +24,112 @@ public class SignHostApiClient
...
+            string checksumHexadecimal = BitConverter
+                .ToString(result)
+                .ToLower()
+                .Replace("-", string.Empty);
+
+            if (checksumHexadecimal != validChecksum) {
+                throw new BadAuthorizationException("Checksum is invalid.");
+            }
+        }
+		*/
```
Comment (WouterHD) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Please don't commit commented code

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 25]
```diff
@@ -283,7 +392,7 @@ public static void RegisterVerification<T>()
 				throw new ArgumentNullException(nameof(filePath));
 			}
 
-			using (Stream fileStream = System.IO.File.Open(
+			using (Stream fileStream = File.Open(
```
Comment (WouterHD) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: This change should not be part of this commit

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 26]
```diff
@@ -21,6 +24,112 @@ public class SignHostApiClient
...
+					using (var sha1 = SHA1.Create())
+					{
+						var preCalculatedChecksum = sha1.ComputeHash(Encoding.UTF8.GetBytes(
+						$"{transactionId}||{status}|{sharedSecret}"));
+						calculatedChecksum = BitConverter.ToString(preCalculatedChecksum).Replace("-", string.Empty).ToLower();
+					};
+				} else {
+					throw new Exception("Invalid postback format.");
+				}
+				}
```
Comment (WouterHD) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Indentation on this closing bracket is not correct

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 27]
```diff
@@ -21,6 +24,112 @@ public class SignHostApiClient
...
+                .ToLower()
+                .Replace("-", string.Empty);
+
+            if (checksumHexadecimal != validChecksum) {
+                throw new BadAuthorizationException("Checksum is invalid.");
+            }
+        }
+		*/
+
+		public static bool IsPostbackChecksumValid(string postback, string sharedSecred)
```
Comment (WouterHD) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Correct spelling of variable to `sharedSecret`

**Comment by WouterHD** [comment]
Names of both the pull request and the commit aren't very descriptive

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 28]
```diff
@@ -26,6 +26,52 @@ public class SignHostApiClient
 		private readonly ISignHostApiClientSettings settings;
 		private readonly HttpClient client;
 
+		public static bool IsPostbackChecksumValid(
+			IDictionary<string, string[]> headers,
+			string body,
+			string sharedSecret,
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: I think I'd rather have the shared secret as part of the ISignhostapiclientsettings ... ?

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 29]
```diff
@@ -26,6 +26,52 @@ public class SignHostApiClient
...
+			string postbackChecksum;
+			string calculatedChecksum;
+			Postback postback;
+
+			try {
+				postback = JsonConvert.DeserializeObject<Postback>(body);
+				postbackTransaction = postback;
+			} catch (InvalidCastException e) {
+				return false;
+			} catch {
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Is there another way to avoid a catch all?

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 30]
```diff
@@ -26,6 +26,52 @@ public class SignHostApiClient
...
+			postbackTransaction = null;
+			string[] postbackChecksumArray;
+			string postbackChecksum;
+			string calculatedChecksum;
+			Postback postback;
+
+			try {
+				postback = JsonConvert.DeserializeObject<Postback>(body);
+				postbackTransaction = postback;
+			} catch (InvalidCastException e) {
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Is this exception raised by json.net?

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 31]
```diff
@@ -26,6 +26,52 @@ public class SignHostApiClient
...
+				return false;
+			} catch {
+				return false;
+			}
+
+			if (headers.ContainsKey("Checksum")) {
+				headers.TryGetValue("Checksum", out postbackChecksumArray);
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: This can be combined into one statement, eg
`if (headers.TryGetValue("Checksum", out postbackChecksumArray)) {` ...
Avoids a double lookup in the dictionary, but makes it more terse / readable

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 32]
```diff
@@ -26,6 +26,52 @@ public class SignHostApiClient
...
+				postbackChecksum = postbackChecksumArray.First();
+			} else {
+				postbackChecksum = postback.Checksum;
+			}
+
+			if (!string.IsNullOrEmpty(postbackTransaction.Id) &&
+				!string.IsNullOrEmpty(postbackChecksum)) {
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: When placing an if statement across multiple lines it often is better readable to place the closing `) {` on a newline.
eg:
```
if (some
   thing
   too long
) {
  // bla bla
}
```

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 33]
```diff
@@ -26,6 +26,52 @@ public class SignHostApiClient
...
+				using (var sha1 = SHA1.Create()) {
+					var preCalculatedChecksum = sha1.ComputeHash(Encoding.UTF8.GetBytes(
+						$"{postbackTransaction.Id}||" +
+						$"{(int)postbackTransaction.Status}|" +
+						$"{sharedSecret}"));
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Reduce the amount of string concatting / building!

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 34]
```diff
@@ -26,6 +26,52 @@ public class SignHostApiClient
...
+					calculatedChecksum = BitConverter.ToString(
+						preCalculatedChecksum)
+ 						.Replace("-", string.Empty)
+						.ToLower();
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: 1. This requires two unneeded string allocations
2. Make it a seperate method for readability. Could be a private method if that makes sense.

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 35]
```diff
@@ -26,6 +26,52 @@ public class SignHostApiClient
...
+				}
+			} else {
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Should be:
```
}
else {
```

File: src/SignhostAPIClient/Rest/SignHostApiClient.cs [hunk 36]
```diff
@@ -26,6 +26,52 @@ public class SignHostApiClient
...
+				return false;
+			}
+			return Equals(calculatedChecksum, postbackChecksum);
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClient.cs]: Always leave a newline after a closing `}` unless it is immediatly followed by another `}`

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 37]
```diff
@@ -15,10 +15,78 @@ public class SignHostApiClientTests
...
...: 203,""Activity"": ""Signed"",""CreatedDateTime"": ""2016-06-15T23:38:04.1965465+02:00""}],""RejectReason"": null,""SignUrl"": ""https://view.signhost.com/sign/d3c93bd6-f1ce-48e7-8c9c-c2babfdd4034"",""SignedDateTime"": null,""RejectDateTime"": null,""CreatedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""ModifiedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""Context"": null}],""Receivers"": [{""Id"": ""97ed6b54-b6d1-46ed-88c1-79779c3b47b1"",""Name"": ""John Doe"",""Email"": ""user@example.com"",""Language"": ""en-US"",""Message"": ""Hello, please find enclosed the digital signed document. Best regards, John Doe"",""Reference"": null,""Activities"": null,""CreatedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""ModifiedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""Context"": null}],""Reference"": ""Contract #123"",""PostbackUrl"": ""https://example.com/postback.php"",""SignRequestMode"": 2,""DaysToExpire"": 30,""SendEmailNotifications"": true,""CreatedDateTime"": ""2016-08-31T21:22:56.2467731+02:00"",""ModifiedDateTime"": ""2016-08-31T21:22:56.2467731+02:00"",""CanceledDateTime"": null,""Context"": null,""Checksum"": ""cdc09eee2ed6df2846dcc193aedfef59f2834f8d""}";
```
Comment (MrJoe) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Might be better to place this in a resx file? Line is way to long

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 38]
```diff
@@ -15,10 +15,78 @@ public class SignHostApiClientTests
...
...: 203,""Activity"": ""Signed"",""CreatedDateTime"": ""2016-06-15T23:38:04.1965465+02:00""}],""RejectReason"": null,""SignUrl"": ""https://view.signhost.com/sign/d3c93bd6-f1ce-48e7-8c9c-c2babfdd4034"",""SignedDateTime"": null,""RejectDateTime"": null,""CreatedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""ModifiedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""Context"": null}],""Receivers"": [{""Id"": ""97ed6b54-b6d1-46ed-88c1-79779c3b47b1"",""Name"": ""John Doe"",""Email"": ""user@example.com"",""Language"": ""en-US"",""Message"": ""Hello, please find enclosed the digital signed document. Best regards, John Doe"",""Reference"": null,""Activities"": null,""CreatedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""ModifiedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""Context"": null}],""Reference"": ""Contract #123"",""PostbackUrl"": ""https://example.com/postback.php"",""SignRequestMode"": 2,""DaysToExpire"": 30,""SendEmailNotifications"": true,""CreatedDateTime"": ""2016-08-31T21:22:56.2467731+02:00"",""ModifiedDateTime"": ""2016-08-31T21:22:56.2467731+02:00"",""CanceledDateTime"": null,""Context"": null,""Checksum"": ""70dda90616f744797972c0d2f787f86643a60c83""}";
```
Comment (MrJoe) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Line too long

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 39]
```diff
@@ -15,10 +15,78 @@ public class SignHostApiClientTests
...
...,{""Id"": ""de94cf6e-e1a3-4c33-93bf-2013b036daaf"",""Code"": 203,""Activity"": ""Signed"",""CreatedDateTime"": ""2016-06-15T23:38:04.1965465+02:00""}],""RejectReason"": null,""SignUrl"": ""https://view.signhost.com/sign/d3c93bd6-f1ce-48e7-8c9c-c2babfdd4034"",""SignedDateTime"": null,""RejectDateTime"": null,""CreatedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""ModifiedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""Context"": null}],""Receivers"": [{""Id"": ""97ed6b54-b6d1-46ed-88c1-79779c3b47b1"",""Name"": ""John Doe"",""Email"": ""user@example.com"",""Language"": ""en-US"",""Message"": ""Hello, please find enclosed the digital signed document. Best regards, John Doe"",""Reference"": null,""Activities"": null,""CreatedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""ModifiedDateTime"": ""2016-06-15T23:33:04.1965465+02:00"",""Context"": null}],""Reference"": ""Contract #123"",""PostbackUrl"": ""https://example.com/postback.php"",""SignRequestMode"": 2,""DaysToExpire"": 30,""SendEmailNotifications"": true,""CreatedDateTime"": ""2016-08-31T21:22:56.2467731+02:00"",""ModifiedDateTime"": ""2016-08-31T21:22:56.2467731+02:00"",""CanceledDateTime"": null,""Context"": null}";
```
Comment (MrJoe) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Line too long

File: src/SignhostAPIClient.Tests/SignHostApiClient.cs [hunk 40]
```diff
@@ -15,10 +15,78 @@ public class SignHostApiClientTests
...
+			result.Should().BeFalse();
+		}
+
+		[Fact]
+		public void when_IsPostbackChecksumValid_is_called_with_valid_postback_in_header_then_true_is_returned()
+		{
+			// Arrange
+			IDictionary<string, string[]> headers = new Dictionary<string, string[]> {
+				{ "Content-Type", new[] { "application/json" }},
+			{"Checksum", new[] {"cdc09eee2ed6df2846dcc193aedfef59f2834f8d"}}
```
Comment (MrJoe) [inline src/SignhostAPIClient.Tests/SignHostApiClient.cs]: Wrong indentation

File: src/SignhostAPIClient/Rest/SignHostApiClientSettings.cs [hunk 41]
```diff
@@ -17,10 +17,19 @@ public SignHostApiClientSettings(string appkey, string apikey)
 			APIKey = apikey;
 		}
 
+		public SignHostApiClientSettings(string appkey, string apikey, string sharedsecret)
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/SignHostApiClientSettings.cs]: 👍 keeps backwards compatibility

File: src/SignhostAPIClient/Rest/ISignhostApiClientSettings.cs [hunk 42]
```diff
@@ -18,6 +18,8 @@ public interface ISignHostApiClientSettings
 		/// </summary>
 		string APPKey { get; }
 
+		string SharedSecret { get; }
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/ISignhostApiClientSettings.cs]: Unfortunately this breaks backward compatibility ... unfortunately we don't have default interface methods (yet), not sure how to fix this without bumping the package

File: src/SignhostAPIClient/Rest/DataObjects/PostbackContext.cs [hunk 43]
```diff
@@ -0,0 +1,18 @@
...
+using System.Globalization;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+
+namespace Signhost.APIClient.Rest.DataObjects
+{
+	public partial class PostbackContext
+	{
+		[JsonProperty("Environment")]
+		public string Environment { get; set; }
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/DataObjects/PostbackContext.cs]: What is this used for?

File: src/SignhostAPIClient/Rest/DataObjects/Link.cs [hunk 44]
```diff
@@ -0,0 +1,16 @@
+﻿using Newtonsoft.Json;
+
+namespace Signhost.APIClient.Rest.DataObjects
+{
+	public partial class Link
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/DataObjects/Link.cs]: Why is this a partial class?

File: src/SignhostAPIClient/Rest/DataObjects/Link.cs [hunk 45]
```diff
@@ -0,0 +1,16 @@
...
+	{
+		[JsonProperty("Rel")]
+		public string Rel { get; set; }
+
+		[JsonProperty("Type")]
+		public string Type { get; set; }
+
+		[JsonProperty("Link")]
+		public string LinkLink { get; set; }
```
Comment (MrJoe) [inline src/SignhostAPIClient/Rest/DataObjects/Link.cs]: I'd rather have the property named Link, perhaps rename the class?

File: src/SignhostAPIClient.Tests/PostbackMock.json [hunk 46]
```diff
@@ -0,0 +1,3 @@
+﻿{
+
+}
```
Comment (MrJoe) [inline src/SignhostAPIClient.Tests/PostbackMock.json]: Remove?

File: src/SignhostAPIClient.Tests/SignhostAPIClient.Tests.csproj [hunk 47]
```diff
@@ -34,6 +27,15 @@
...
-</Project>
\ No newline at end of file
+  <ItemGroup>
+    <Compile Condition=" '$(EnableDefaultCompileItems)' == 'true' " Update="RequestBodies.Designer.cs">
+      <DependentUpon>RequestBodies.resx</DependentUpon>
+    </Compile>
+  </ItemGroup>
+  <ItemGroup>
+    <None Remove="PostbackMock.json" />
+  </ItemGroup>
```
Comment (MrJoe) [inline src/SignhostAPIClient.Tests/SignhostAPIClient.Tests.csproj]: Remove?

File: src/SignhostAPIClient/SignhostAPIClient.csproj [hunk 48]
```diff
@@ -1,6 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup Label="Build">
-    <TargetFrameworks>netstandard1.4;net45</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.0</TargetFrameworks>
```
Comment (MrJoe) [inline src/SignhostAPIClient/SignhostAPIClient.csproj]: Should be a seperate commit

File: src/SignhostAPIClient/SignhostAPIClient.csproj [hunk 49]
```diff
@@ -48,4 +50,7 @@
   <ItemGroup Label=".NET standard Package References" Condition="'$(TargetFramework)' == 'netstandard1.4'">
     <PackageReference Include="System.Security.Cryptography.Algorithms" Version="4.3.*" />
   </ItemGroup>
+  <ItemGroup>
+    <Compile Remove="Rest\DataObjects\Link.cs" />
+  </ItemGroup>
```
Comment (MrJoe) [inline src/SignhostAPIClient/SignhostAPIClient.csproj]: Remove

File: src/SignhostAPIClient/SignhostAPIClient.csproj [hunk 50]
```diff
@@ -40,6 +40,8 @@
     <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
     <PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta001" PrivateAssets="All" />
     <PackageReference Include="System.ValueTuple" Version="4.3.1" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.7.0" />
```
Comment (MrJoe) [inline src/SignhostAPIClient/SignhostAPIClient.csproj]: Remove, should only be in the Test project
